sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -256,6 +278,12 @@ static void _on_resp_timeout(void *arg) {
 #endif
         event_timeout_set(&memo->resp_evt_tmout, timeout);
 
+        if (memo->state == GCOAP_MEMO_WAIT) {
```

Comment:
The issue with disarming timers is that that'd open up to a race condition described in https://github.com/RIOT-OS/RIOT/issues/14390 -- that condition already exists, but I'd prefer not to add another case.

I've toyed around a bit with such solutions, but they were invariably more complex than just accepting that a timer fires a few more times than absolutely necessary without doing anything than decrement a counter.

Yes, there's some random component (and one might consider trimming that off the decrement-only case); the way it is phrased now, the CON timeout will happen just when it would have happened if there had not been any ACK. (Conceptually, the application and retransmission timeouts are different things anyway, but it's a practical choice widespread in implementations to conflate them. Thus, sticking with what would happen if there was no ACK is consistent.).

Reply:
I mean you could just not add the random jitter in the `memo->state == GCOAP_MEMO_WAIT` case, if that is desirable.

Reply:
Could, but I doubt it'd make it better. The backup calculation code is already hard to read as it was, I don't expect that skipping the jitter will make the ROM size shrink, and at runtime it'd save the random number generation of an unlikely event at cost of another jump.

When the general raciness is addressed, there'll be a code path that explicitly calculates the remaining timeout and that can then be done bypassing the jitter as it happens at a completely different point, but that's for #14390.